tests/unit/fem/test_pa_grad.cpp

```diff
@@ -0,0 +1,254 @@
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "catch.hpp"
+#include "mfem.hpp"
+
+using namespace mfem;
+
+double compare_pa_assembly(int dim, int num_elements, int order, bool transpose)
+{
+   Mesh * mesh;
+   if (num_elements == 0)
+   {
+      if (dim == 2)
+      {
+         mesh = new Mesh("../../data/star.mesh", order);
+      }
+      else
+      {
+         mesh = new Mesh("../../data/beam-hex.mesh", order);
+      }
+   }
+   else
+   {
+      if (dim == 2)
+      {
+         mesh = new Mesh(num_elements, num_elements, Element::QUADRILATERAL, true);
+      }
+      else
+      {
+         mesh = new Mesh(num_elements, num_elements, num_elements,
+                         Element::HEXAHEDRON, true);
+      }
+   }
+   FiniteElementCollection *h1_fec = new H1_FECollection(order, dim);
+   FiniteElementCollection *nd_fec = new ND_FECollection(order, dim);
+   FiniteElementSpace h1_fespace(mesh, h1_fec);
+   FiniteElementSpace nd_fespace(mesh, nd_fec);
+
+   DiscreteLinearOperator assembled_grad(&h1_fespace, &nd_fespace);
+   assembled_grad.AddDomainInterpolator(new GradientInterpolator);
+   const int skip_zeros = 1;
+   assembled_grad.Assemble(skip_zeros);
+   assembled_grad.Finalize(skip_zeros);
+   const SparseMatrix& assembled_grad_mat = assembled_grad.SpMat();
+
+   DiscreteLinearOperator pa_grad(&h1_fespace, &nd_fespace);
+   pa_grad.SetAssemblyLevel(AssemblyLevel::PARTIAL);
+   pa_grad.AddDomainInterpolator(new GradientInterpolator);
+   pa_grad.Assemble();
+   pa_grad.Finalize();
+
+   int insize, outsize;
+   if (transpose)
+   {
+      insize = nd_fespace.GetVSize();
+      outsize = h1_fespace.GetVSize();
+   }
+   else
+   {
+      insize = h1_fespace.GetVSize();
+      outsize = nd_fespace.GetVSize();
+   }
+   Vector xv(insize);
+   Vector assembled_y(outsize);
+   Vector pa_y(outsize);
+
+   xv.Randomize();
+   if (transpose)
+   {
+      assembled_grad_mat.BuildTranspose();
+      assembled_grad_mat.MultTranspose(xv, assembled_y);
+      pa_grad.MultTranspose(xv, pa_y);
+   }
+   else
+   {
+      assembled_grad_mat.Mult(xv, assembled_y);
+      pa_grad.Mult(xv, pa_y);
+   }
+
+   if (false)
+   {
+      std::cout << "true   \tpa\n";
+      for (int i = 0; i < assembled_y.Size(); ++i)
+      {
+         std::cout << i << " : " << assembled_y(i) << "\t" << pa_y(i) << std::endl;
+      }
+   }
+
+   pa_y -= assembled_y;
+   double error = pa_y.Norml2() / assembled_y.Norml2();
+   std::cout << "dim " << dim << " ne " << num_elements << " order "
+             << order;
+   if (transpose)
+   {
+      std::cout << " T";
+   }
+   std::cout << ": error in PA gradient: " << error << std::endl;
```

Comment:
See my other comment in `test_pa_idinterp.cpp` about writing to stdout in tests.